Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder.py

```diff
@@ -0,0 +1,614 @@
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1.  Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+# 2.  Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import fnmatch
+import itertools
+import json
+import re
+import sys
+import urllib
+from collections import OrderedDict
+
+from webkitpy.layout_tests.controllers.test_result_writer import TestResultWriter
+from webkitpy.layout_tests.models.test import Test
+from webkitpy.w3c.common import TEMPLATED_TEST_HEADER
+
+supported_test_extensions = (
+    ".htm",
+    ".html",
+    ".mht",
+    ".php",
+    ".pl",
+    ".py",
+    ".shtml",
+    ".svg",
+    ".xht",
+    ".xhtml",
+    ".xml",
+)
+assert len(supported_test_extensions) == len(set(supported_test_extensions))
+assert list(supported_test_extensions) == sorted(supported_test_extensions)
+
+supported_reference_extensions = (
+    ".htm",
+    ".html",
+    ".svg",
+    ".xht",
+    ".xhtml",
+    ".xml",
+)
+assert len(supported_reference_extensions) == len(set(supported_reference_extensions))
+assert list(supported_reference_extensions) == sorted(supported_reference_extensions)
+assert set(supported_reference_extensions) < set(supported_test_extensions)
+
+
+skipped_directories = {
+    ".svn",
+    "_svn",
+    "resources",
+    "support",
+    "script-tests",
+    "tools",
+    "reference",
+    "reftest",
+}
+
+# Files ending in -ref/-notref aren't considered expectations, but are skipped.
+skipped_test_suffixes = tuple(
+    "".join(pair)
+    for pair in itertools.product(
+        ("-expected", "-expected-mismatch", "-ref", "-notref"),
+        supported_reference_extensions,
+    )
+)
+
+digit_re = re.compile("([0-9]+)")
+
+
+def natsort(string_to_split):
+    split = digit_re.split(string_to_split)
+    # this converts the split numbers into tuples so that "01" < "1"
```

Comment:
Probably should be `"1" < "02"` as your example, right?

Reply:
No; the goal here is to provide a strict total ordering of strings, which means we need to be able to distinguish "01" and "1".